README.md
# phonegap-build-api-js [![Build Status][travis-ci-img]][travis-ci-url]

> Node.js REST Client for the PhoneGap Build API

## Overview

This library simplfies authentication and requests to the
[PhoneGap Build REST API][build-api-docs] for node.js clients.

In many ways, this library is little more than a convenience wrapper
for [mikeal's][github-mikeal] [request][github-request] library. You can expect
that all of [request's][github-request] functionality to be available to the
`API` object returned by `client.auth();`.

If something is inaccurate or missing, please send a pull request!

## Usage

### Authenticate with Username and Password

    var client = require('phonegap-build-api');

    client.auth({ username: 'zelda', password: 'tr1f0rce' }, function(e, api) {
        // time to make requests
    });

### Authenticate with Token

    var client = require('phonegap-build-api');

    client.auth({ token: 'abc123' }, function(e, api) {
        // time to make requests
    });

### GET https://build.phonegap.com/api/v1/me

    api.get('/me', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### GET https://build.phonegap.com/api/v1/apps

    api.get('/apps', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### GET https://build.phonegap.com/api/v1/apps/:id

    api.get('/apps/199692', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### GET https://build.phonegap.com/api/v1/apps/:id/icon

    api.get('/apps/199692/icon').pipe(fs.createWriteStream('icon.png'));

### GET https://build.phonegap.com/api/v1/apps/:id/:platform

    api.get('/apps/199692/android').pipe(fs.createWriteStream('app.apk'));

### GET https://build.phonegap.com/api/v1/keys

    api.get('/keys', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### GET https://build.phonegap.com/api/v1/keys/:platform

    api.get('/keys/ios', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### GET https://build.phonegap.com/api/v1/keys/:platform/:id

    api.get('/keys/ios/917', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/apps

    var options = {
        form: {
            data: {
                title: 'My App',
                create_method: 'file'
            },
            file: '/path/to/app.zip'
        }
    };

    api.post('/apps', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### PUT https://build.phonegap.com/api/v1/apps/:id

    var options = {
        form: {
            data: {
                debug: false
            },
            file: '/path/to/app.zip'
        }
    };

    api.put('/apps/197196', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/apps/:id/icon

    var options = {
        form: {
            icon: 'my-icon.png'
        }
    };

    api.post('/apps/232741/icon', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/apps/:id/build

Build all platforms:

    api.post('/apps/232741/build', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });
 
Build specific platforms:

    var options = {
        form: {
            data: {
                platforms: [ 'android', 'blackberry', 'ios', 'winphone', 'webos' ]
            }
        }
    };

    api.post('/apps/232741/build', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/apps/:id/build/:platform

    api.post('/apps/232741/build/android', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/apps/:id/collaborators

    var options = {
        form: {
            data: {
                email: 'michael@michaelbrooks.ca',
                role: 'dev'
            }
        }
    };

    api.post('/apps/232741/collaborators', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### PUT https://build.phonegap.com/api/v1/apps/:id/collaborators/:id

    var options = {
        form: {
            data: {
                role: 'tester'
            }
        }
    };

    api.put('/apps/232741/collaborators/263955', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### POST https://build.phonegap.com/api/v1/keys/:platform

    var options = {
        form: {
            data: {
                title: 'My BlackBerry Signing Key',
                password: 'my-password'
            },
            db: '/path/to/sigtool.db',
            csk: '/path/to/sigtool.csk'
        }
    };

    api.post('/keys/blackberry', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### PUT https://build.phonegap.com/api/v1/keys/:platform/:id

    var options = {
        form: {
            data: {
                password: 'my-updated-password'
            }
        }
    };

    api.put('/keys/blackberry/1505', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### DELETE https://build.phonegap.com/api/v1/apps/:id

    api.del('/apps/14450', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### DELETE https://build.phonegap.com/api/v1/apps/:id/collaborators/:id

    api.del('/apps/232741/collaborators/263955', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### DELETE https://build.phonegap.com/api/v1/keys/:platform/:id

    api.del('/keys/ios/2729', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

## API

### client.auth(options, callback)

PhoneGap Build Authentication.

Authentications with PhoneGap Build and returns an instance of `API`.
The authentication credentials can be a username and password or user-token.

#### Options:

  - `options` `{Object}` is the authentication settings.
  - `options.username` `{String}` is the phonegap build username.
  - `options.password` `{String}` is the phonegap build password.
  - `options.token` `{String}` can be used instead of username and password.
  - [`options.proxy`] `{String}` specifies an optional proxy server. e.g. 'http://myproxy.com:8181'.
  - `callback` `{Function}` is trigger after the authentication.
    - `e` `{Error}` is null unless there is an error.
    - `api` `{Object}` is the `API` instance to interact with phonegap build.

#### Example:

    var client = require('phonegap-build-api');

    client.auth({ username: 'zelda', password: 'tr1force' }, function(e, api) {
        if (e) {
            console.log('error:', e);
            return;
        }

        // make some api requests
    });

### api(path, [options], [callback])

API Request.

Create a RESTful request to the PhoneGap Build API. The `api` function is a
wrapper to [request][github-request]'s interface.

The `path` parameter is a relative path to a PhoneGap Build API response.
For example, to the resource `https://build.phonegap.com/api/v1/me` is specified
as the path `/me`.

The `options` parameter maps directly to [request][github-request]'s options.

The default request method is `GET`. You can specify a specific but you can be changed
in the `options` parameters (e.g. `{ method: 'POST' }`).

To send form data, you can use the `options.form` parameter. The key `data` is
assumed to be JSON and all other keys are assumed to be file paths.

#### Options:

  - `path` `{String}` is a relative resource path (e.g. `"/apps"`).
  - `[options]` `{Object}` is a request options object.
  - `[callback]` `{Function}` is trigger after the request
    - `e` `{Error}` is null unless there is an error
    - `data` `{Object}` is the JSON response.

#### Example: GET Request

    api('/me', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

#### Example: POST Request

    var options = {
        form: {
            data: {
                title: 'My App',
                create_method: 'file'
            },
            file: '/path/to/app.zip'
        },
        method: 'POST'
    };

    api('/apps', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### api.get(path, [options], [callback])

GET API Request.

A convenience function for `api(path, [options], [callback])`, where `options`
uses `{ method: 'GET' }`.

#### Options:

  - `path` `{String}` is a relative resource path (e.g. `"/apps"`).
  - `[options]` `{Object}` is a request options object.
  - `[callback]` `{Function}` is trigger after the request
    - `e` `{Error}` is null unless there is an error
    - `data` `{Object}` is the JSON response.

#### Example:

    api.get('/me', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### api.post(path, [options], [callback])

POST API Request.

A convenience function for `api(path, [options], [callback])`, where `options`
uses `{ method: 'POST' }`.

#### Options:

  - `path` `{String}` is a relative resource path (e.g. `"/apps"`).
  - `[options]` `{Object}` is a request options object.
  - `[callback]` `{Function}` is trigger after the request
    - `e` `{Error}` is null unless there is an error
    - `data` `{Object}` is the JSON response.

#### Example:

    var options = {
        form: {
            data: {
                title: 'My App',
                create_method: 'file'
            },
            file: '/path/to/app.zip'
        }
    };

    api.post('/apps', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### api.put(path, [options], [callback])

PUT API Request.

A convenience function for `api(path, [options], [callback])`, where `options`
uses `{ method: 'PUT' }`.

#### Options:

  - `path` `{String}` is a relative resource path (e.g. `"/apps"`).
  - `[options]` `{Object}` is a request options object.
  - `[callback]` `{Function}` is trigger after the request
    - `e` `{Error}` is null unless there is an error
    - `data` `{Object}` is the JSON response.

#### Example:

    var options = {
        form: {
            data: {
                debug: false
            },
            file: '/path/to/app.zip'
        }
    };

    api.put('/apps/197196', options, function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### api.del(path, [options], [callback])

DELETE API Request.

A convenience function for `api(path, [options], [callback])`, where `options`
uses `{ method: 'DELETE' }`.

#### Options:

  - `path` `{String}` is a relative resource path (e.g. `"/apps"`).
  - `[options]` `{Object}` is a request options object.
  - `[callback]` `{Function}` is trigger after the request
    - `e` `{Error}` is null unless there is an error
    - `data` `{Object}` is the JSON response.

#### Example:

    api.del('/apps/14450', function(e, data) {
        console.log('error:', e);
        console.log('data:', data);
    });

### api.defaults(options)

This maps directly to [request][github-request]'s `default` method.

> This method returns a wrapper around the normal request API that defaults to whatever options you pass in to it.

## Alternative Libraries

### Java

- [pgbuild-api][pgbuild-api] by [Hardeep Shoker][github-hardeep]

### Node.js

- [phonegapbuildapi][github-phonegapbuildapi] by [germallon][github-germallon]

[travis-ci-img]: https://travis-ci.org/mwbrooks/phonegap-build-api-js.png?branch=master
[travis-ci-url]: http://travis-ci.org/mwbrooks/phonegap-build-api-js
[build-api-docs]: https://build.phonegap.com/docs/api
[github-mikeal]: https://github.com/mikeal
[github-request]: https://github.com/mikeal/request
[pgbuild-api]: https://github.com/hardeep/pgbuild-api
[github-hardeep]: https://github.com/hardeep
[github-phonegapbuildapi]: https://github.com/germallon/phonegapbuildapi
[github-germallon]: https://github.com/germallon



CHANGELOG.md
0.3.2 / 2012-01-25
==================

  * Add proxy support.

0.3.1 / 2012-01-10
==================

  * Add usage example for token authentication.
  * Rename project from 'phonegap-build-rest-js' to 'phonegap-build-api-js'.
  * Unpublish NPM modules 'phonegap-build-rest-js'.
  * Publish NPM module 'phonegap-build-api-js'.
  
0.3.0 / 2012-01-06
==================

  * Fix issue where empty error would cause failure.
  * Add multipart/form-data support.
  * Add file upload support.
  * Support response codes 201 (created) and 202 (accepted).
  * Update README.md with usage examples.

0.2.0 / 2012-12-05
==================

  * Expose `API` as a public object.
  * NPM no longer registers as a preferred global.

0.1.0 / 2012-11-06
==================

  * Initial release


lib/api.js
/*
 * Module dependencies.
 */

var request = require('./request'),
    Arguments = require('./arguments'),
    fs = require('fs'),
    p = require('path');

/**
 * API Request Constructor.
 *
 * Creating an instance of the API Request will enable the RESTful access
 * to a PhoneGap Build API.
 *
 * Options:
 *
 *   - `options` {Object} defines the API server that has been authenticated.
 *   - `api.token` {String} is the user token for authentication.
 *   - `api.protocol` {String} is the protocol, e.g. 'http' or 'https'.
 *   - `api.host` {String} is the host address, e.g. 'build.phonegap.com'.
 *   - `api.port` {String} is the port, e.g. '80' or '443'.
 *   - `api.path` {String} is the api path, e.g. '/api/v1'
 *   - [`api.path`] {String} is a proxy to forward requests, e.g. 'http://myproxy.com'
 */

module.exports = function(options) {
    // require options argument
    if (!options) throw new Error('missing options argument');
    if (!options.token) throw new Error('missing options.token argument');
    if (!options.protocol) throw new Error('missing options.protocol argument');
    if (!options.host) throw new Error('missing options.host argument');
    if (!options.port) throw new Error('missing options.port argument');
    if (!options.path) throw new Error('missing options.path argument');

    /**
     * API Request.
     *
     * Sends an HTTP request to the PhoneGap Build API.
     *
     * By default, the request method is `GET` but can be changed
     * in the `options` parameters (e.g. `{ method: 'POST' }`).
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    var api = function(path, options, callback) {
        var args = new Arguments(path, options, callback);
        args.options.method = args.options.method || 'GET';
        return api.request(args.path, args.options, args.callback);
    };

    /**
     * GET API Request.
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    api.get = function(path, options, callback) {
        var args = new Arguments(path, options, callback);
        args.options.method = 'GET';
        return api.request(args.path, args.options, args.callback);
    };

    /**
     * POST API Request.
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    api.post = function(path, options, callback) {
        var args = new Arguments(path, options, callback);
        args.options.method = 'POST';
        return api.request(args.path, args.options, args.callback);
    };

    /**
     * PUT API Request.
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    api.put = function(path, options, callback) {
        var args = new Arguments(path, options, callback);
        args.options.method = 'PUT';
        return api.request(args.path, args.options, args.callback);
    };

    /**
     * DELETE API Request.
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    api.del = function(path, options, callback) {
        var args = new Arguments(path, options, callback);
        args.options.method = 'DELETE';
        return api.request(args.path, args.options, args.callback);
    };

    /**
     * API Request Defaults.
     *
     * The method simply forwards to [requests](https://github.com/mikeal/request)
     * default method.
     */

    api.defaults = function(options, requester) {
        return request.defaults(options, requester);
    };

    /*
     * API Configuration.
     *
     * Internal configuration values for constructing an API request.
     *
     * Values:
     *
     *   - `api.token` {String} is the user token for authentication.
     *   - `api.protocol` {String} is the protocol, e.g. 'http' or 'https'.
     *   - `api.host` {String} is the host address, e.g. 'build.phonegap.com'.
     *   - `api.port` {String} is the port, e.g. '80' or '443'.
     *   - `api.path` {String} is the api path, e.g. '/api/v1'
     *   - [`api.proxy`] {String} is proxy for requests through, e.g. 'http://myproxy.com'
     */

    api.token = options.token;
    api.protocol = options.protocol;
    api.host = options.host;
    api.port = options.port;
    api.path = options.path;
    api.defaults.version = options.version;
    api.proxy = options.proxy;

    /*
     * API Request Handler.
     *
     * Private method that handles all types of API requests.
     *
     * Options:
     *
     *   - `path` {String} is a resource path (e.g. `"/apps"`).
     *   - `[options]` {Object} is a request options object.
     *   - `[callback]` {Function} is trigger after the request
     *     - `e` {Error} is null unless there is an error
     *     - `data` {Object} is the JSON response.
     */

    api.request = function(path, options, callback) {
        // require arguments
        if (!path) throw new Error('missing path argument');
        if (!options) throw new Error('missing options argument');

        // optional arguments
        callback = callback || function() {};

        var args = new Arguments(path, callback);
        var uri = api.protocol + '//' +
                  api.host + ':' +
                  api.port +
                  api.path +
                  args.path + '?auth_token=' + api.token;

        // hijack the form option.
        // this is a clean interface to define form data, but the
        // phonegap build api does not support encoded querystring data.
        var data = options.form;
        options.form = undefined;

        // use proxy when provided
        options.proxy = api.proxy;

        var r = request.send(uri, options, function(e, res, body) {
            if (e) {
                // error in request
                args.callback(e);
            }
            else if (res.statusCode !== 200 && res.statusCode !== 201 && res.statusCode !== 202) {
                // eror in response
                args.callback(new Error(body));
            }
            else {
                try {
                    body = JSON.parse(body);
                }
                catch(err) {
                    // continue but data untouched (likely image data)
                }

                if (typeof body.error === 'string') {
                    // api response includes an error
                    args.callback(new Error(body.error));
                }
                else {
                    // api response is successful
                    args.callback(null, body);
                }
            }
        });

        // form-data support as content-type multipart/form-data
        if (data) {
            var form = r.form(),
                key,
                value;

            // handle each key in the form object:
            //   - `data` {Object} is handled as JSON.
            //   - `*` {String} is handled as a file path.
            for(key in data) {
                if (key === 'data') {
                    value = JSON.stringify(data[key]);
                }
                else {
                    value = fs.createReadStream(p.normalize(data[key]));
                }

                form.append(key, value);
            }
        }

        return r;
    };

    return api;
};

module.exports.prototype = {
};
